File: leetcode/901.股票价格跨度.go
type StockSpanner struct {
    stack [][2]int
    n int
}


func Constructor() StockSpanner {
    return StockSpanner{}
}


func (s *StockSpanner) Next(price int) int {
    // 单调栈
    s.n += 1
    ans := -1
    for len(s.stack) > 0 && s.stack[len(s.stack) - 1][1] <= price {
        s.stack = s.stack[:len(s.stack) - 1]
    }
    if len(s.stack) != 0 {
        ans = s.n - s.stack[len(s.stack) - 1][0]
    } else {
        ans = s.n
    }
    s.stack = append(s.stack, [2]int{s.n, price})
    return ans
}


/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */

File: leetcode/901.股票价格跨度.py
class StockSpanner:

    def __init__(self):
        self.stack = []
        self.n = 0
    def next(self, price: int) -> int:
        self.n += 1
        ans = -1
        while len(self.stack) and self.stack[-1][0] <= price:
            self.stack.pop()
        if len(self.stack):
            ans = self.n - self.stack[-1][1]
        else:
            ans = self.n
        self.stack.append((price, self.n))
        return ans

        
        


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)